velox/functions/lib/TimeUtils.h

```diff
@@ -104,4 +106,82 @@ struct InitSessionTimezone {
     timeZone_ = getTimeZoneFromConfig(config);
   }
 };
+
+/// Return day-of-year (DOY) of the first `firstDayOfWeek` in the year.
+/// If the `firstDayOfWeek` is Monday, it returns DOY of first Monday in
+/// the year. The returned DOY is a number from 1 to 7.
+///
+/// `firstDayOfWeek` is a 1-based weekday number starting with Sunday.
+///   (1 = Sunday, 2 = Monday, ..., 7 = Saturday).
+FOLLY_ALWAYS_INLINE
+uint32_t getDayOfFirstDayOfWeek(int32_t y, uint32_t firstDayOfWeek) {
+  auto firstDay =
+      date::year_month_day(date::year(y), date::month(1), date::day(1));
+  auto weekday = date::weekday(firstDay).c_encoding() + 1;
+
+  int32_t delta = firstDayOfWeek - weekday;
+  if (delta < 0) {
+    delta += 7;
+  }
+
+  return delta + 1;
+}
+
+/// Return week-based-year of year month day.
+/// The week containing Jan 1 has `minimalDaysInFirstWeek` or more days is
+/// week 1. The week is defined via `firstDayOfWeek`. `firstDayOfWeek` is a
+/// 1-based weekday number starting with Sunday.
+///   (1 = Sunday, 2 = Monday, ..., 7 = Saturday).
+///
+/// For ISO 8601, `firstDayOfWeek` is 2 (Monday) and `minimalDaysInFirstWeek`
+/// is 4. For Legacy Spark, `firstDayOfWeek` is 1 (Sunday) and
+/// `minimalDaysInFirstWeek` is 1 by default.
+///
+/// getWeekYear only works with gregorian calendar due to limitations in the
+/// date library. As a result, dates before the gregorian calendar was used
+/// (1582-10-15) would yield mismatched results.
+///
+/// The algorithm refers to the weekyear algorithm in jdk:
+/// https://github.com/openjdk/jdk8/blob/6a383433a9f4661a96a90b2a4c7b5b9a85720031/jdk/src/share/classes/java/util/GregorianCalendar.java#L2077
```

Comment:
This link points to an implementation in JDK 8. Is there any difference in the implementations among JDK versions?

Reply:
There's no difference. I will updated the URL to the most recent openjdk. https://github.com/openjdk/jdk/blob/d9c67443f7d7f03efb2837b63ee2acc6113f737f/src/java.base/share/classes/java/util/GregorianCalendar.java#L2058